common/txmgr/address_state.go

```diff
@@ -325,31 +325,37 @@ func (as *addressState[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) moveIn
 	return nil
 }
 
-// moveUnconfirmedToConfirmed moves the unconfirmed transaction to the confirmed state.
-// It returns an error if there is no unconfirmed transaction with the given ID.
+// moveTxToConfirmed moves a transaction to the confirmed state.
+// It returns an error if there is no transaction with the given ID.
 // It returns an error if there is no attempt with the given receipt.
-func (as *addressState[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) moveUnconfirmedToConfirmed(
+func (as *addressState[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) moveTxToConfirmed(
 	receipt txmgrtypes.ChainReceipt[TX_HASH, BLOCK_HASH],
 ) error {
 	as.Lock()
 	defer as.Unlock()
 
 	txAttempt, ok := as.attemptHashToTxAttempt[receipt.GetTxHash()]
 	if !ok {
-		return fmt.Errorf("move_unconfirmed_to_confirmed: no attempt found with receipt %v", receipt)
+		return fmt.Errorf("no attempt found with receipt %q", receipt.GetTxHash())
 	}
 
 	txAttempt.Receipts = []txmgrtypes.ChainReceipt[TX_HASH, BLOCK_HASH]{receipt}
-	tx, ok := as.unconfirmedTxs[txAttempt.TxID]
+	tx, ok := as.allTxs[txAttempt.TxID]
 	if !ok {
-		return fmt.Errorf("move_unconfirmed_to_confirmed: no unconfirmed transaction with ID %d", txAttempt.TxID)
+		return fmt.Errorf("no transaction with ID %d found for hash %q", txAttempt.TxID, receipt.GetTxHash())
 	}
 	txAttempt.State = txmgrtypes.TxAttemptBroadcast
 	if txAttempt.BroadcastBeforeBlockNum == nil {
 		blockNum := receipt.GetBlockNumber().Int64()
 		txAttempt.BroadcastBeforeBlockNum = &blockNum
 	}
 	tx.State = TxConfirmed
```

Comment:
If we're marking the tx as confirmed here, wouldn't we need to move it from the map that it's in to the confirmedTxs map?

Reply:
Yeah I think we really need to optimize everything into a single map to help avoid issues such as this. Made a ticket: https://smartcontract-it.atlassian.net/browse/BCI-2869. Either way we need to add a check here to move the state in the map.